src/linear_algebra/clifford_algebra/fold.lean

```diff
@@ -0,0 +1,148 @@
+/-
+Copyright (c) 2022 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import linear_algebra.clifford_algebra.conjugation
+
+/-!
+# Recursive computation rules for the Clifford algebra
+
+This file provides API for a special case `clifford_algebra.foldr` of the universal property
+`clifford_algebra.lift` with `A = module.End R N` for some arbitrary module `N`. This specialization
+resembles the `list.foldr` operation, allowing a bilinear map to be "folded" along the generators.
+
+For convenience, this file also provides `clifford_algebra.foldl`, implemented via
+`clifford_algebra.reverse`
+
+## Main definitions
+
+* `clifford_algebra.foldr`: a computation rule for building linear maps out of the clifford
+  algebra starting on the right, analogous to using `list.foldr` on the generators.
+* `clifford_algebra.foldl`: a computation rule for building linear maps out of the clifford
+  algebra starting on the left, analogous to using `list.foldl` on the generators.
+
+## Main statements
+
+* `clifford_algebra.right_induction`: an induction rule that adds generators from the right.
+* `clifford_algebra.left_induction`: an induction rule that adds generators from the left.
+-/
+
+universes u1 u2 u3
+
+variables {R M N : Type*}
+variables [comm_ring R] [add_comm_group M] [add_comm_group N]
+variables [module R M] [module R N]
+variables (Q : quadratic_form R M)
+
+namespace clifford_algebra
+
+section foldr
+
+/-- Fold a bilinear map along the generators of a term of the clifford algebra, with the rule
+given by `foldr Q f hf n (ι Q m * x) = f m (foldr Q f hf n x)`.
+
+For example, `foldr f hf n (r • ι R u + ι R v * ι R w) = r • f u n + f v (f w n)`. -/
+def foldr (f : M →ₗ[R] N →ₗ[R] N) (hf : ∀ m x, f m (f m x) = Q m • x) :
```

Comment:
I'm really struggling to get on board with the names `foldl` and `foldr`.

I think my main issue is that their naming makes sense only with interpretation that terms of the Clifford algebra are (sums of) products of elements of `M` which is valid of course but not canonical.

Reply:
Would you prefer `foldr_generators` or similar as a name?